crates/ra_hir_expand/src/builtin_macro.rs

```diff
@@ -1,28 +1,76 @@
 //! Builtin macro
 use crate::db::AstDatabase;
+use crate::quote;
 use crate::{
     ast::{self, AstNode},
-    name, AstId, CrateId, HirFileId, MacroCallId, MacroDefId, MacroDefKind, MacroFileKind,
-    TextUnit,
+    eager::EagerResult,
+    name,
+    util::unquote,
+    AstId, CrateId, HirFileId, MacroCallId, MacroDefId, MacroDefKind, MacroFileKind, TextUnit,
 };
-
-use crate::quote;
+use once_cell::sync::Lazy;
+use ra_parser::FragmentKind;
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub enum BuiltinExpander {
     Line,
+    Concat,
+}
+
+struct BuiltInMacroInfo {
+    name: name::Name,
+    kind: BuiltinExpander,
+    expand: fn(
+        db: &dyn AstDatabase,
+        id: MacroCallId,
+        _tt: &tt::Subtree,
+    ) -> Result<tt::Subtree, mbe::ExpandError>,
+    eager: Option<fn(&tt::Subtree) -> Result<EagerResult, mbe::ExpandError>>,
+}
+
+impl BuiltInMacroInfo {
+    fn new(
+        name: name::Name,
+        kind: BuiltinExpander,
+        expand: fn(
+            db: &dyn AstDatabase,
+            id: MacroCallId,
+            _tt: &tt::Subtree,
+        ) -> Result<tt::Subtree, mbe::ExpandError>,
+        eager: Option<fn(&tt::Subtree) -> Result<EagerResult, mbe::ExpandError>>,
+    ) -> Self {
+        BuiltInMacroInfo { name, kind, expand, eager }
+    }
 }
 
+const BUILTIN_MACROS: Lazy<Vec<BuiltInMacroInfo>> = Lazy::new(|| {
```

Comment:
Can this be just a const array instead of going via lazy ?

Reply:
const fn has been stabilized.

Reply:
I tried it before but this error occurred: 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ca46ba3b0a8bf3a13ef4513362e0dc8f

Maybe I should setup the array without new method ?

Reply:
Ah, yeah, indeed it's only the `new` method that is problematic: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=91450be14655a138183db9529bc7b133

Yeah, I think I prefer to not use lazy static, even if that means typing the struct literal